visualisation_temporal.py

```diff
@@ -0,0 +1,118 @@
+# packages
+import h5py
+import json
+import os
+import pandas as pd
+import pickle
+import plotly.io as pio
+import plotnine as p9
+import numpy as np
+import random
+
+
+
+###################################################
+# USER INPUT: config to use
+
+
+# CONFIG_JSON = "config/simulation-charmander.json"
+# CONFIG_JSON = "config/simulation-charmeleon.json"
+# CONFIG_JSON = "config/simulation-charizard.json"
+# CONFIG_JSON = "config/simulation-charmeleon-temporal.json"
+CONFIG_JSON = "config/debugging-measurement-times.json"
+
+###################################################
+
+
+
+
+# constants
+with open(CONFIG_JSON, "r") as file:
+    CONFIG = json.load(file)
+SIM_DIR = f"out/{CONFIG['simulation_name']}/simulation/remaster"
+SIM_PICKLE_DIR = f"out/{CONFIG['simulation_name']}/simulation/pickle"
+DB_PATH = f"out/{CONFIG['simulation_name']}/{CONFIG['output_hdf5']}"
+PLOT_DIR = f"out/{CONFIG['simulation_name']}/plots"
+if not os.path.exists(PLOT_DIR):
+    os.makedirs(PLOT_DIR)
+    
+MAX_NUM_SIMS_TO_PLOT = 50
+
+
+
+
+
+# Read all temporal data out
+
+def _read_temporal_data(key, db_conn):
+    return {
+        "key": key,
+        "key_num": int(key.split("_")[1]),
+        "temporal_data": db_conn[f"{key}/output/parameters/temporal_measurements"][()],
+    }
+    
+DB_CONN = h5py.File(DB_PATH, "r")
+all_temporal_data = [
+    _read_temporal_data(key, DB_CONN) for key in DB_CONN.keys() if key.startswith("record")
+]
+DB_CONN.close()
+
+
+# Pack a sample of simulations into a dataframe in order to plot
+inds_of_sims_to_plot = random.sample(range(len(all_temporal_data)),
+        (MAX_NUM_SIMS_TO_PLOT if len(all_temporal_data) > MAX_NUM_SIMS_TO_PLOT else len(all_temporal_data)))
+temporal_data_sample_df = pd.concat([pd.DataFrame(
```

Comment:
Here is a simpler way to construct the `temporal_data_sample_df` which avoid reading all of the data

```python
def _read_temporal_data(key, db_conn):
    df = pd.DataFrame(db_conn[f"{key}/output/parameters/temporal_measurements"][()])
    df["key"] = key
    return df

DB_CONN = h5py.File(DB_PATH, "r")
keys_list = list(DB_CONN.keys())
num_samples = MAX_NUM_SIMS_TO_PLOT if len(keys_list) > MAX_NUM_SIMS_TO_PLOT else len(keys_list)
temporal_data_sample_df = pd.concat([_read_temporal_data(key, DB_CONN) for key in random.sample(keys_list, num_samples)])
DB_CONN.close()
```

Reply:
Wow, yep definitely a lot simpler. That's my bad, I was pretty much following your workflow from the <code>visualisation.py</code> script, I should have simplified it down. Have fixed now.